File: src/world/World.ts
import type { GameConfig } from "../engine/core/Config.js";
import { ChunkSlotStatus, SharedPool } from "../engine/alloc/SharedPool.js";
import type { WorkerHandle } from "../engine/workers/WorkerSpawner.js";
import type { MeshDoneMessage, WorldGenDoneMessage, WorldGenJobMessage, MeshJobMessage, WorkerOutboundMessage } from "../engine/workers/messages.js";
import type { SaveManager } from "../engine/save/SaveManager.js";
import { BlockRegistry } from "./BlockRegistry.js";
import { Chunk } from "./Chunk.js";
import { ChunkManager, packChunkKey } from "./ChunkManager.js";

const worldToChunk = (coordinate: number, size: number): number => Math.floor(coordinate / size);
const mod = (value: number, size: number): number => ((value % size) + size) % size;
const chunkSeed = (chunkX: number, chunkZ: number, seed: number): number => {
  let h = seed ^ Math.imul(chunkX, 374761393) ^ Math.imul(chunkZ, 668265263);
  h = Math.imul(h ^ (h >>> 13), 1274126177);
  return h >>> 0;
};

export interface WorldBlockRef {
  readonly chunk: Chunk;
  readonly localX: number;
  readonly localZ: number;
  readonly index: number;
}

export class World {
  private readonly chunks = new ChunkManager();
  private readonly slotToChunk = new Map<number, Chunk>();
  private readonly pendingGen: Chunk[] = [];
  private readonly pendingMesh: Chunk[] = [];
  private saveManager: SaveManager | null = null;
  private centerChunkX = Number.NaN;
  private centerChunkZ = Number.NaN;

  constructor(
    private readonly pool: SharedPool,
    private readonly worldGenWorkers: WorkerHandle[],
    private readonly mesherWorkers: WorkerHandle[],
    private readonly blocks: BlockRegistry,
    private readonly config: GameConfig,
  ) {
    for (const handle of this.worldGenWorkers) {
      handle.worker.onmessage = (event: MessageEvent<WorkerOutboundMessage>) => {
        const message = event.data;
        if (message.kind === "generated") this.onWorldGenDone(handle, message);
      };
    }

    for (const handle of this.mesherWorkers) {
      handle.worker.onmessage = (event: MessageEvent<WorkerOutboundMessage>) => {
        const message = event.data;
        if (message.kind === "meshed") this.onMeshDone(handle, message);
      };
    }
  }

  update(cameraPosition: ArrayLike<number>): void {
    const chunkX = worldToChunk(cameraPosition[0], this.config.chunkSize);
    const chunkZ = worldToChunk(cameraPosition[2], this.config.chunkSize);

    this.centerChunkX = chunkX;
    this.centerChunkZ = chunkZ;
    this.ensureVisibleRadius(chunkX, chunkZ);
    this.unloadFarChunks(chunkX, chunkZ);

    this.pumpQueues();
  }

  dispose(): void {
    for (const handle of this.worldGenWorkers) handle.worker.terminate();
    for (const handle of this.mesherWorkers) handle.worker.terminate();
  }

  attachSaveManager(saveManager: SaveManager): void {
    this.saveManager = saveManager;
  }

  isReady(): boolean {
    if (!Number.isFinite(this.centerChunkX) || !Number.isFinite(this.centerChunkZ)) return false;
    const center = this.chunks.get(this.centerChunkX, this.centerChunkZ);
    return Boolean(center && (center.state === "meshReady" || center.state === "uploaded"));
  }

  hasChunkKey(key: string): boolean {
    return this.chunks.hasKey(key);
  }

  *entries(): IterableIterator<[string, Chunk]> {
    for (const chunk of this.chunks.values()) {
      yield [chunk.key, chunk];
    }
  }

  getChunkSlot(chunk: Chunk) {
    return this.pool.view(chunk.slotIndex);
  }

  getChunk(chunkX: number, chunkZ: number): Chunk | undefined {
    return this.chunks.get(chunkX, chunkZ);
  }

  getChunkBySlotIndex(slotIndex: number): Chunk | undefined {
    return this.slotToChunk.get(slotIndex);
  }

  getBlockIdAt(worldX: number, worldY: number, worldZ: number): number {
    if (worldY < 0 || worldY >= this.config.worldHeight) return 0;
    return this.getBlockId(worldX, worldY, worldZ);
  }

  getRedstonePackedAt(worldX: number, worldY: number, worldZ: number): number {
    const ref = this.resolveBlock(worldX, worldY, worldZ);
    if (!ref) return 0;
    return this.pool.view(ref.chunk.slotIndex).redstone[ref.index];
  }

  resolveBlock(worldX: number, worldY: number, worldZ: number): WorldBlockRef | null {
    if (worldY < 0 || worldY >= this.config.worldHeight) return null;
    const chunkX = worldToChunk(worldX, this.config.chunkSize);
    const chunkZ = worldToChunk(worldZ, this.config.chunkSize);
    const chunk = this.chunks.get(chunkX, chunkZ);
    if (!chunk) return null;
    const localX = mod(worldX, this.config.chunkSize);
    const localZ = mod(worldZ, this.config.chunkSize);
    const index = (worldY * this.config.chunkSize + localZ) * this.config.chunkSize + localX;
    return { chunk, localX, localZ, index };
  }

  setBlockIdAt(worldX: number, worldY: number, worldZ: number, blockId: number): boolean {
    const ref = this.resolveBlock(worldX, worldY, worldZ);
    if (!ref) return false;
    const slot = this.pool.view(ref.chunk.slotIndex);
    if (slot.voxels[ref.index] === blockId) return false;
    slot.voxels[ref.index] = blockId;
    if (blockId === 0) slot.redstone[ref.index] = 0;
    this.markChunkDirty(ref.chunk.chunkX, ref.chunk.chunkZ);
    this.requestRemesh(ref.chunk);
    return true;
  }

  setRedstonePackedAt(worldX: number, worldY: number, worldZ: number, packed: number): boolean {
    const ref = this.resolveBlock(worldX, worldY, worldZ);
    if (!ref) return false;
    const slot = this.pool.view(ref.chunk.slotIndex);
    if (slot.redstone[ref.index] === packed) return false;
    slot.redstone[ref.index] = packed;
    this.markChunkDirty(ref.chunk.chunkX, ref.chunk.chunkZ);
    this.requestRemesh(ref.chunk);
    return true;
  }

  markChunkDirty(chunkX: number, chunkZ: number): void {
    this.saveManager?.markDirty(chunkX, chunkZ);
  }

  requestRemesh(chunk: Chunk): void {
    if (chunk.state === "loadingFromDisk") return;
    if (chunk.state === "queuedMesh") return;
    if (chunk.state === "meshing") {
      chunk.needsRemesh = true;
      return;
    }
    if (
      chunk.state === "generating" ||
      chunk.state === "queuedGen" ||
      chunk.state === "voxelsReady"
    ) {
      return;
    }
    chunk.state = "queuedMesh";
    this.pendingMesh.push(chunk);
  }

  markUploaded(chunk: Chunk): void {
    chunk.state = "uploaded";
    const slot = this.pool.view(chunk.slotIndex);
    Atomics.store(slot.status, 0, ChunkSlotStatus.GPU_UPLOADED);
  }

  isSolid(worldX: number, worldY: number, worldZ: number): boolean {
    if (worldY < 0 || worldY >= this.config.worldHeight) return false;
    const blockId = this.getBlockId(worldX, worldY, worldZ);
    if (blockId === 0) return false;
    const block = this.blocks.tryGet(blockId);
    if (!block) return false;
    const { collision } = block;
    return collision.maxX > collision.minX && collision.maxY > collision.minY && collision.maxZ > collision.minZ;
  }

  isFluid(worldX: number, worldY: number, worldZ: number): boolean {
    if (worldY < 0 || worldY >= this.config.worldHeight) return false;
    const blockId = this.getBlockId(worldX, worldY, worldZ);
    if (blockId === 0) return false;
    const block = this.blocks.tryGet(blockId);
    return Boolean(block?.material.liquid);
  }

  onSaveLoadSuccess(
    chunkX: number,
    chunkZ: number,
    voxels: Uint8Array,
    light: Uint8Array,
    redstone: Uint8Array,
  ): void {
    const chunk = this.chunks.get(chunkX, chunkZ);
    if (!chunk) return;
    const slot = this.pool.view(chunk.slotIndex);
    Atomics.store(slot.chunkX, 0, chunkX);
    Atomics.store(slot.chunkZ, 0, chunkZ);
    slot.voxels.set(voxels);
    slot.light.set(light);
    slot.redstone.set(redstone);
    Atomics.store(slot.status, 0, ChunkSlotStatus.VOXELS_READY);
    chunk.state = "voxelsReady";
    chunk.needsRemesh = false;
    this.pendingMesh.push(chunk);
  }

  onSaveLoadFailed(chunkX: number, chunkZ: number): void {
    const chunk = this.chunks.get(chunkX, chunkZ);
    if (!chunk) return;
    chunk.state = "queuedGen";
    this.pendingGen.push(chunk);
  }

  private ensureVisibleRadius(centerChunkX: number, centerChunkZ: number): void {
    for (let dz = -this.config.renderDistance; dz <= this.config.renderDistance; dz++) {
      for (let dx = -this.config.renderDistance; dx <= this.config.renderDistance; dx++) {
        const chunkX = centerChunkX + dx;
        const chunkZ = centerChunkZ + dz;
        if (this.chunks.has(chunkX, chunkZ)) continue;

        const slot = this.pool.acquire();
        if (!slot) return;

        const chunk = new Chunk(chunkX, chunkZ, slot.slotIndex);
        Atomics.store(slot.chunkX, 0, chunkX);
        Atomics.store(slot.chunkZ, 0, chunkZ);
        this.chunks.set(chunk);
        this.slotToChunk.set(chunk.slotIndex, chunk);
        if (this.saveManager) {
          chunk.state = "loadingFromDisk";
          this.saveManager.requestLoad(chunkX, chunkZ);
        } else {
          chunk.state = "queuedGen";
          this.pendingGen.push(chunk);
        }
      }
    }
  }

  private unloadFarChunks(centerChunkX: number, centerChunkZ: number): void {
    const maxDistance = this.config.renderDistance + 1;
    const toUnload: Chunk[] = [];

    for (const chunk of this.chunks.values()) {
      const dx = Math.abs(chunk.chunkX - centerChunkX);
      const dz = Math.abs(chunk.chunkZ - centerChunkZ);
      const canRelease =
        chunk.state === "meshReady" ||
        chunk.state === "uploaded" ||
        chunk.state === "meshFailed";

      if ((dx > maxDistance || dz > maxDistance) && canRelease) {
        toUnload.push(chunk);
      }
    }

    for (const chunk of toUnload) {
      this.slotToChunk.delete(chunk.slotIndex);
      this.chunks.delete(chunk.chunkX, chunk.chunkZ);
      this.pool.release(this.pool.view(chunk.slotIndex));
    }
  }

  private pumpQueues(): void {
    for (const handle of this.worldGenWorkers) {
      if (handle.busy) continue;
      const chunk = this.pendingGen.shift();
      if (!chunk || this.slotToChunk.get(chunk.slotIndex) !== chunk) continue;
      const slot = this.pool.view(chunk.slotIndex);
      Atomics.store(slot.status, 0, ChunkSlotStatus.GENERATING);
      chunk.state = "generating";
      handle.busy = true;
      const message: WorldGenJobMessage = {
        kind: "generate",
        slotIndex: chunk.slotIndex,
        chunkX: chunk.chunkX,
        chunkZ: chunk.chunkZ,
        seed: chunkSeed(chunk.chunkX, chunk.chunkZ, this.config.worldSeed),
      };
      handle.worker.postMessage(message);
    }

    for (const handle of this.mesherWorkers) {
      if (handle.busy) continue;
      const chunk = this.pendingMesh.shift();
      if (!chunk || this.slotToChunk.get(chunk.slotIndex) !== chunk) continue;
      const slot = this.pool.view(chunk.slotIndex);
      Atomics.store(slot.status, 0, ChunkSlotStatus.MESHING);
      chunk.state = "meshing";
      handle.busy = true;
      const message: MeshJobMessage = {
        kind: "mesh",
        slotIndex: chunk.slotIndex,
      };
      handle.worker.postMessage(message);
    }
  }

  private onWorldGenDone(handle: WorkerHandle, message: WorldGenDoneMessage): void {
    handle.busy = false;
    const chunk = this.slotToChunk.get(message.slotIndex);
    if (!chunk) return;
    chunk.state = "voxelsReady";
    chunk.needsRemesh = false;
    this.saveManager?.markDirty(chunk.chunkX, chunk.chunkZ);
    this.pendingMesh.push(chunk);
  }

  private onMeshDone(handle: WorkerHandle, message: MeshDoneMessage): void {
    handle.busy = false;
    const chunk = this.slotToChunk.get(message.slotIndex);
    if (!chunk) return;
    chunk.vertexCount = message.vertexCount;
    chunk.indexCount = message.indexCount;
    chunk.state = message.success ? "meshReady" : "meshFailed";
    if (chunk.needsRemesh) {
      chunk.needsRemesh = false;
      this.requestRemesh(chunk);
    }
  }

  private getBlockId(worldX: number, worldY: number, worldZ: number): number {
    const chunkX = worldToChunk(worldX, this.config.chunkSize);
    const chunkZ = worldToChunk(worldZ, this.config.chunkSize);
    const chunk = this.chunks.get(chunkX, chunkZ);
    if (!chunk) return 0;

    const slot = this.pool.view(chunk.slotIndex);
    const localX = mod(worldX, this.config.chunkSize);
    const localZ = mod(worldZ, this.config.chunkSize);
    const index = (worldY * this.config.chunkSize + localZ) * this.config.chunkSize + localX;
    return slot.voxels[index];
  }
}

File: src/world/World.cpp
#include "World.hpp"
#include "ChunkCoords.hpp"
#include <algorithm>
#include <cmath>
#include <cstring>

namespace terrain {
namespace {

inline auto redstoneIndex(int32_t worldY, int32_t localX, int32_t localZ, int32_t chunkSize) -> int32_t {
  return (worldY * chunkSize + localZ) * chunkSize + localX;
}

} // namespace

World::World(SharedPool& pool,
             const GameConfig& config,
             IChunkWorker& worker,
             IChunkPersistence* persistence)
  : m_pool(pool), m_config(config),
    m_jobQueue(worker),
    m_persistence(persistence)
{}

void World::update(glm::vec3 cameraPos) {
  int32_t cx = worldToChunk(cameraPos.x, m_config.chunkSize);
  int32_t cz = worldToChunk(cameraPos.z, m_config.chunkSize);

  m_centerChunkX = cx;
  m_centerChunkZ = cz;
  m_hasCenter = true;

  ensureVisibleRadius(cx, cz);
  unloadFarChunks(cx, cz);
  m_jobQueue.pump(m_chunks, m_pool, m_slotToChunk, m_config.worldSeed);
  m_remeshScheduler.flush([this](int32_t chunkX, int32_t chunkZ) {
    auto* chunk = m_chunks.getMut(chunkX, chunkZ);
    if (!chunk) return;
    requestBoundaryNeighborRemeshes(*chunk, 0, 0);
    requestBoundaryNeighborRemeshes(*chunk, m_config.chunkSize - 1, m_config.chunkSize - 1);
  });
}

auto World::isReady() const -> bool {
  if (!m_hasCenter) return false;
  auto* center = m_chunks.get(m_centerChunkX, m_centerChunkZ);
  return center && chunkHasMeshes(*center);
}

auto World::hasTerrain() const -> bool {
  if (!m_hasCenter) return false;
  auto* center = m_chunks.get(m_centerChunkX, m_centerChunkZ);
  return center && chunkHasTerrainData(*center);
}

auto World::getChunk(int32_t cx, int32_t cz) const -> const Chunk* {
  return m_chunks.get(cx, cz);
}

auto World::getChunkMut(int32_t cx, int32_t cz) -> Chunk* {
  return m_chunks.getMut(cx, cz);
}

auto World::getChunkBySlotIndex(int32_t slotIndex) const -> const Chunk* {
  auto it = m_slotToChunk.find(slotIndex);
  if (it == m_slotToChunk.end()) return nullptr;
  return m_chunks.get(it->second.cx, it->second.cz);
}

auto World::getChunkBySlotIndex(int32_t slotIndex) -> Chunk* {
  auto it = m_slotToChunk.find(slotIndex);
  if (it == m_slotToChunk.end()) return nullptr;
  return m_chunks.getMut(it->second.cx, it->second.cz);
}

auto World::getChunkSlot(const Chunk& chunk) -> ChunkSlot {
  return m_pool.view(chunk.slotIndex);
}

auto World::getRedstonePackedAt(int32_t worldX, int32_t worldY, int32_t worldZ) const -> uint8_t {
  if (worldY < 0 || worldY >= m_config.worldHeight) return 0;
  int32_t cx = floorToChunk(worldX, m_config.chunkSize);
  int32_t cz = floorToChunk(worldZ, m_config.chunkSize);
  const auto* chunk = m_chunks.get(cx, cz);
  if (!chunk) return 0;
  int32_t localX = mod(worldX, m_config.chunkSize);
  int32_t localZ = mod(worldZ, m_config.chunkSize);
  auto slot = m_pool.view(chunk->slotIndex);
  return slot.redstone[redstoneIndex(worldY, localX, localZ, m_config.chunkSize)];
}

auto World::setRedstonePackedAt(int32_t worldX, int32_t worldY, int32_t worldZ, uint8_t packed) -> bool {
  if (worldY < 0 || worldY >= m_config.worldHeight) return false;
  int32_t cx = floorToChunk(worldX, m_config.chunkSize);
  int32_t cz = floorToChunk(worldZ, m_config.chunkSize);
  auto* chunk = m_chunks.getMut(cx, cz);
  if (!chunk) return false;

  int32_t localX = mod(worldX, m_config.chunkSize);
  int32_t localZ = mod(worldZ, m_config.chunkSize);
  auto slot = m_pool.view(chunk->slotIndex);
  int32_t idx = redstoneIndex(worldY, localX, localZ, m_config.chunkSize);

  if (slot.redstone[idx] == packed) return false;
  slot.redstone[idx] = packed;
  markChunkDirty(chunk->chunkX, chunk->chunkZ);
  requestRemesh(*chunk);

  // Defer boundary neighbor remeshes
  if (localX == 0 || localX == m_config.chunkSize - 1 ||
      localZ == 0 || localZ == m_config.chunkSize - 1) {
    m_remeshScheduler.noteBoundaryEdit(chunk->chunkX, chunk->chunkZ);
  }
  return true;
}

void World::markUploaded(const Chunk& chunk) {
  auto* mutChunk = m_chunks.getMut(chunk.chunkX, chunk.chunkZ);
  if (mutChunk) mutChunk->state = ChunkState::Uploaded;
  auto slot = m_pool.view(chunk.slotIndex);
  *slot.status = static_cast<int32_t>(ChunkSlotStatus::GPU_UPLOADED);
}

void World::clear() {
  std::vector<int32_t> usedSlots;
  m_chunks.forEach([&](const Chunk& chunk) { usedSlots.push_back(chunk.slotIndex); });

  for (const int32_t slotIndex : usedSlots) {
    m_slotToChunk.erase(slotIndex);
    m_pool.release(m_pool.view(slotIndex));
  }

  m_chunks = ChunkManager{};
  m_slotToChunk.clear();
  m_jobQueue.clear();
  m_pendingUploadSlots.clear();
  m_remeshScheduler.clear();
  m_hasCenter = false;
}

void World::onWorldGenDone(int32_t slotIndex) {
  auto* chunk = getChunkBySlotIndex(slotIndex);
  if (!chunk) return;
  if (chunk->state != ChunkState::Generating) return;

  chunk->state = ChunkState::DensityReady;
  chunk->needsRemesh = false;
  markChunkDirty(chunk->chunkX, chunk->chunkZ);
  m_jobQueue.pushMesh(chunk->slotIndex, chunk->chunkX, chunk->chunkZ);
  requestNeighborRemeshes(*chunk);
}

void World::onMeshDone(int32_t slotIndex, uint32_t vertexCount, uint32_t indexCount, bool success) {
  auto* chunk = getChunkBySlotIndex(slotIndex);
  if (!chunk) return;
  if (chunk->state != ChunkState::Meshing) return;
  chunk->vertexCount = vertexCount;
  chunk->indexCount = indexCount;
  if (success) {
    chunk->meshRestartRetries = 0;
    chunk->state = ChunkState::MeshReady;
    m_pendingUploadSlots.push_back(slotIndex);
  } else if (chunk->meshRestartRetries < MAX_CHUNK_MESH_RESTART_RETRIES) {
    ++chunk->meshRestartRetries;
    restartChunkFromScratch(*chunk);
    return;
  } else {
    chunk->state = ChunkState::MeshFailed;
  }
  if (chunk->needsRemesh) {
    chunk->needsRemesh = false;
    requestRemesh(*chunk);
  }
}

auto World::drainPendingUploadSlots() -> std::vector<int32_t> {
  return std::move(m_pendingUploadSlots);
}

void World::onSaveLoadFailed(int32_t chunkX, int32_t chunkZ) {
  auto* chunk = m_chunks.getMut(chunkX, chunkZ);
  if (!chunk) return;
  if (chunk->state != ChunkState::LoadingFromDisk) return;
  chunk->state = ChunkState::QueuedGen;
  m_jobQueue.pushGen(chunk->slotIndex, chunk->chunkX, chunk->chunkZ);
}

// ---- Private ----

void World::ensureVisibleRadius(int32_t centerCX, int32_t centerCZ) {
  int32_t r = m_config.renderDistance;
  int32_t r2 = r * r;
  struct ChunkOffset {
    int32_t dx;
    int32_t dz;
  };

  std::vector<ChunkOffset> visibleOffsets;
  visibleOffsets.reserve(static_cast<size_t>((r * 2 + 1) * (r * 2 + 1)));

  for (int32_t dz = -r; dz <= r; ++dz) {
    for (int32_t dx = -r; dx <= r; ++dx) {
      if (dx * dx + dz * dz > r2) continue;
      visibleOffsets.push_back({dx, dz});
    }
  }

  std::sort(visibleOffsets.begin(), visibleOffsets.end(),
            [](const ChunkOffset& a, const ChunkOffset& b) {
              const int32_t distA = a.dx * a.dx + a.dz * a.dz;
              const int32_t distB = b.dx * b.dx + b.dz * b.dz;
              if (distA != distB) return distA < distB;

              const int32_t manhattanA = std::abs(a.dx) + std::abs(a.dz);
              const int32_t manhattanB = std::abs(b.dx) + std::abs(b.dz);
              if (manhattanA != manhattanB) return manhattanA < manhattanB;

              if (a.dz != b.dz) return a.dz < b.dz;
              return a.dx < b.dx;
            });

  for (const ChunkOffset& offset : visibleOffsets) {
    int32_t cx = centerCX + offset.dx;
    int32_t cz = centerCZ + offset.dz;
    if (m_chunks.has(cx, cz)) continue;

    auto slot = m_pool.acquire();
    if (!slot) return;

    Chunk chunk{cx, cz, slot->slotIndex};
    *slot->chunkX = cx;
    *slot->chunkZ = cz;
    m_chunks.set(chunk);

    auto* inserted = m_chunks.getMut(cx, cz);
    if (inserted) {
      m_slotToChunk[chunk.slotIndex] = {cx, cz};
    }

    if (m_persistence) {
      inserted->state = ChunkState::LoadingFromDisk;
      m_persistence->requestLoad(cx, cz);
    } else {
      inserted->state = ChunkState::QueuedGen;
      m_jobQueue.pushGen(inserted->slotIndex, inserted->chunkX, inserted->chunkZ);
    }
  }
}

void World::unloadFarChunks(int32_t centerCX, int32_t centerCZ) {
  int32_t maxDist = m_config.renderDistance + 1;
  int32_t maxDist2 = maxDist * maxDist;
  std::vector<Chunk*> toUnload;

  m_chunks.forEach([&](const Chunk& chunk) {
    int32_t dx = std::abs(chunk.chunkX - centerCX);
    int32_t dz = std::abs(chunk.chunkZ - centerCZ);
    bool canRelease =
      chunk.state == ChunkState::MeshReady ||
      chunk.state == ChunkState::Uploaded ||
      chunk.state == ChunkState::MeshFailed;
    if (dx * dx + dz * dz > maxDist2 && canRelease) {
      toUnload.push_back(const_cast<Chunk*>(&chunk));
    }
  });

  for (auto* chunk : toUnload) {
    m_slotToChunk.erase(chunk->slotIndex);
    m_chunks.remove(chunk->chunkX, chunk->chunkZ);
    m_pool.release(m_pool.view(chunk->slotIndex));
  }
}

void World::restartChunkFromScratch(Chunk& chunk) {
  auto slot = m_pool.view(chunk.slotIndex);
  *slot.vertexCount = 0u;
  *slot.indexCount = 0u;
  *slot.renderFlags = 0u;
  *slot.opaqueIndexCount = 0u;
  *slot.transparentIndexCount = 0u;
  chunk.vertexCount = 0u;
  chunk.indexCount = 0u;
  chunk.opaqueIndexCount = 0u;
  chunk.transparentIndexCount = 0u;
  chunk.hasOpaque = false;
  chunk.hasTransparent = false;
  chunk.needsRemesh = false;

  if (m_persistence) {
    chunk.state = ChunkState::LoadingFromDisk;
    m_persistence->requestLoad(chunk.chunkX, chunk.chunkZ);
  } else {
    chunk.state = ChunkState::QueuedGen;
    m_jobQueue.pushGen(chunk.slotIndex, chunk.chunkX, chunk.chunkZ);
  }
}

void World::requestRemesh(Chunk& chunk) {
  if (chunk.state == ChunkState::LoadingFromDisk) return;
  if (chunk.state == ChunkState::QueuedMesh) return;
  if (chunk.state == ChunkState::Meshing) {
    chunk.needsRemesh = true;
    return;
  }
  if (chunk.state == ChunkState::Generating ||
      chunk.state == ChunkState::QueuedGen ||
      chunk.state == ChunkState::DensityReady) {
    return;
  }
  chunk.state = ChunkState::QueuedMesh;
  m_jobQueue.pushMesh(chunk.slotIndex, chunk.chunkX, chunk.chunkZ);
}

void World::markChunkDirty(int32_t cx, int32_t cz) {
  if (m_persistence) m_persistence->markDirty(cx, cz);
}

auto World::chunkHasTerrainData(const Chunk& chunk) const -> bool {
  return chunk.state == ChunkState::DensityReady ||
         chunk.state == ChunkState::QueuedMesh ||
         chunk.state == ChunkState::Meshing ||
         chunk.state == ChunkState::MeshReady ||
         chunk.state == ChunkState::Uploaded ||
         chunk.state == ChunkState::MeshFailed;
}

auto World::chunkHasMeshes(const Chunk& chunk) const -> bool {
  return chunk.state == ChunkState::MeshReady ||
         chunk.state == ChunkState::Uploaded;
}

void World::requestNeighborRemeshes(const Chunk& chunk) {
  requestNeighborRemesh(chunk, -1, 0);
  requestNeighborRemesh(chunk, 1, 0);
  requestNeighborRemesh(chunk, 0, -1);
  requestNeighborRemesh(chunk, 0, 1);
}

void World::requestBoundaryNeighborRemeshes(const Chunk& chunk, int32_t localX, int32_t localZ) {
  if (localX == 0) requestNeighborRemesh(chunk, -1, 0);
  if (localX == m_config.chunkSize - 1) requestNeighborRemesh(chunk, 1, 0);
  if (localZ == 0) requestNeighborRemesh(chunk, 0, -1);
  if (localZ == m_config.chunkSize - 1) requestNeighborRemesh(chunk, 0, 1);
}

void World::requestNeighborRemesh(const Chunk& chunk, int32_t dx, int32_t dz) {
  auto* neighbor = m_chunks.getMut(chunk.chunkX + dx, chunk.chunkZ + dz);
  if (!neighbor || !chunkHasTerrainData(*neighbor)) return;
  requestRemesh(*neighbor);
}

} // namespace terrain
